src/core/sandbox.c

```diff
@@ -0,0 +1,801 @@
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(eventfd2), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* rt_sigprocmask */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(rt_sigprocmask), 3,
+		SCMP_A0(SCMP_CMP_EQ, SIG_UNBLOCK),
+		SCMP_A2(SCMP_CMP_EQ, NULL),
+		SCMP_A3(SCMP_CMP_EQ, 8));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(rt_sigprocmask), 1,
+		SCMP_A0(SCMP_CMP_EQ, SIG_SETMASK));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(rt_sigprocmask), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* rt_sigaction */
+	const int sigaction_signal_array[13] = {
+		SIGALRM,
+		SIGCHLD,
+		SIGCONT,
+		SIGFPE,
+		SIGHUP,
+		SIGINT,
+		SIGPIPE,
+		SIGQUIT,
+		SIGTERM,
+		SIGTRAP,
+		SIGUSR1,
+		SIGUSR2,
+		SIGWINCH
+	};
+	for (i = 0; i < 13; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(rt_sigaction), 2,
+			SCMP_A0(SCMP_CMP_EQ, sigaction_signal_array[i]),
+			SCMP_A3(SCMP_CMP_EQ, 8));
+		if (rc < 0)
+			goto fail;
+	}
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(rt_sigaction), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* sendto */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(sendto), 4,
+		SCMP_A0(SCMP_CMP_GE, 4),
+		SCMP_A3(SCMP_CMP_EQ, 0),
+		SCMP_A4(SCMP_CMP_EQ, NULL),
+		SCMP_A5(SCMP_CMP_EQ, 0));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(sendto), 70);
+	if (rc < 0)
+		goto fail;
+
+	/* recvfrom */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(recvfrom), 5,
+		SCMP_A0(SCMP_CMP_GE, 4),
+		SCMP_A2(SCMP_CMP_EQ, 8),
+		SCMP_A3(SCMP_CMP_EQ, 0),
+		SCMP_A4(SCMP_CMP_EQ, NULL),
+		SCMP_A5(SCMP_CMP_EQ, NULL));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(recvfrom), 70);
+	if (rc < 0)
+		goto fail;
+
+	/* access */
+	const int access_mode_array[3] = {
+		R_OK,
+		F_OK,
+		X_OK
+	};
+	for (i = 0; i < 3; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(access), 1,
+			SCMP_A1(SCMP_CMP_EQ, access_mode_array[i]));
+		if (rc < 0)
+			goto fail;
+	}
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(access), 60);
+	if (rc < 0)
+		goto fail;
+
+	/* write */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(write), 2,
+		SCMP_A0(SCMP_CMP_LT, 3),
+		SCMP_A2(SCMP_CMP_LE, 4096));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(write), 2,
+		SCMP_A0(SCMP_CMP_GT, 3),
+		SCMP_A2(SCMP_CMP_LE, 4096));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(write), 2,
+		SCMP_A0(SCMP_CMP_EQ, 3),
+		/* SCMP_A1(SCMP_CMP_EQ, "\1\0\0\0\0\0\0\0"), */
+		SCMP_A2(SCMP_CMP_EQ, 8));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(write), 200);
+	if (rc < 0)
+		goto fail;
+
+	/* read */
+	struct read_struct_t {
+		int fd;
+		size_t count;
+	};
+	const struct read_struct_t read_struct[6] = {
+		{ 0, 256  },
+		{ 3, 16   },
+		{ 3, 512  },
+		{ 3, 832  },
+		{ 3, 4000 },
+		{ 3, 4096 }
```

Comment:
This seems a bit too system-specific for my taste; it depends on the buffering algorithm used to read/write, amongst other things.

Since `close()` is whitelisted (we could limit it to file descriptors `>4` perhaps?), we cannot rely on these integers to carry any significance.

If there is no solid reasoning for these numbers, would you consider changing it to some sort of "sane upper bound" instead? (ie ALLOW if it's `<=16384`)